temfpy/optimization.py

```diff
@@ -1,24 +1,125 @@
+"""Test capabilities for optimization.
+
+This module contains a host of models and functions often used for testing optimization algorithms.
+
+"""
 import numpy as np
-import math
 
 
-def ackley(x, a=20, b=0.2, c=2 * math.pi):
+def ackley(x, a=20, b=0.2, c=2 * np.pi):
     r"""Ackley function.
+
+    :math:`f(x) = -a exp(-b \sqrt{\frac{1}{d} \sum_{i=1}^d x_i^2})
+    - exp(\frac{1}{d} \sum_{i=1}^d cos(c x_i)) + a + exp(1)`
+
+    Parameters
+    ----------
+
+    x : array_like
+        Input domain with dimension d.
+        It is usually evaluated on the hypercube
+        :math:`x_i\in [-32.768, 32.768]`, for all :math:`i = 1, \dots, d`.
+
+    a : float, optional
+        The default value is 20
+
+    b : float, optional
+        The default value is 0.2
+
+    c : float, optional
+        The default value is 2π
+
+    Returns
+    -------
+
+    float
+         Output domain
+
+    Notes
+    -----
+    This function was proposed by David Ackley in [A1987]_.
+    It is characterized by an almost flat outer region and a central hole or peak
+    where modulations become more and more influential. The function has
+    its global minimum :math:`f(x) = 0` at :math:`x = (0, \dots, 0)`.
+
+    References
+    ----------
+
+    .. [A1987] Ackley, D. H. (1987).
+       A connectionist machine for genetic hillclimbing.
+       Boston, MA: Kluwer Academic Publishers.
+
+    .. [B1996] Back, T. (1996).
+       Evolutionary algorithms in theory and practice:
+       Evolution strategies, evolutionary programming, genetic algorithms.
+       Oxford, UK: Oxford University Press.
+
+    .. [M2005] Molga, M., and Smutnicki, C. (2005).
+       Test functions for optimization needs.
+       Retrieved June 2020, from http://www.zsd.ict.pwr.wroc.pl/files/docs/functions.pdf.
+
+    Examples
+    --------
+
+    >>> x = [0, 0]
+    >>> y = ackley(x)
+    >>> np.testing.assert_almost_equal(y, 0)
+
     """
-    n = len(x)
 
-    rslt = a + math.exp(1) - (a * (math.exp(-b * math.sqrt(1 / n * np.sum(x ** 2)))))
-    rslt -= math.exp(1 / n * np.sum(np.cos(c * x)))
+    rslt = (
+        a + np.exp(1) - (a * (np.exp(-b * np.sqrt(1 / len(x) * np.sum(np.square(x))))))
+    )
+    rslt -= np.exp(1 / len(x) * np.sum(np.cos(np.multiply(c, x))))
 
     return rslt
 
 
 def rastrigin(x, a=10):
     r"""Rastrigin function.
+
+    :math:`f(x) = a d + \sum_{i=1}^n (x_i^2 -10cos(2\pi x_i))`

```

Comment:
The index of the sum should run to `d`, right? Also, please add a space between the 10 and cos.